python/ray/data/impl/compute.py

```diff
@@ -47,8 +47,23 @@ def apply(self, fn: Any, remote_args: dict,
         ]
         new_blocks, new_metadata = zip(*refs)
 
-        map_bar.block_until_complete(list(new_blocks))
-        new_metadata = ray.get(list(new_metadata))
+        new_metadata = list(new_metadata)
+        try:
+            new_metadata = map_bar.fetch_until_complete(new_metadata)
+        except (ray.exceptions.RayTaskError, KeyboardInterrupt) as e:
+            # One or more mapper tasks failed, or we received a SIGINT signal
+            # while waiting; either way, we cancel all map tasks.
+            for ref in new_metadata:
+                ray.cancel(ref)
+            # Wait until all tasks have failed or been cancelled.
+            for ref in new_metadata:
+                try:
+                    ray.get(ref)
```

Comment:
How about we add a small timeout here, to avoid blocking?

Reply:
Hmm if we do that, aren't we no longer guaranteed to wait until all tasks have been cancelled? I.e. couldn't we end up returning to the console with tasks still outstanding and sending logs to the driver in that case?

Reply:
This also isn't a critical path so I don't immediately see the harm in strict blocking here. Provided that the object will eventually be marked with an error (which I think we guarantee?), this will eventually return for each object and the original exception will be reraised.